### ISSUE TITLE
Speed Is All You Need: On-Device Acceleration of Large Diffusion Models via GPU-Aware Optimizations

### ISSUE THREAD

**Issue by eehoeskrap** [eehoeskrap/PaperReview#19]
Paper : https://arxiv.org/abs/2304.11267
Blog : https://ai.googleblog.com/2023/06/speed-is-all-you-need-on-device.html

이번에 리뷰할 논문은 Google LLC에서 제안된 **Speed Is All You Need: On-Device Acceleration of Large Diffusion Models via GPU-Aware Optimizations** 입니다. On-device 환경에서 Diffusion 모델을 배포하기 위한 방법을 담고있습니다. Large diffusion model은 보통 10억 개 이상의 매개변수를 갖고 있어서 제한된 computing 능력, 메모리 리소스 등 문제가 있다고 하네요. 본 논문에서는 GPU 환경에서 Stable Diffusion 1.4 모델을 약 20회 정도 반복하여 추론 했을 때 512x512 이미지에 대해 INT8 quantization 없이 12초 미만으로 추론해내는 결과를 달성했다고 합니다. 

## GPU-Aware Optimizations

본 논문에서 주로 다루는 내용은 large diffusion model을 사용하여 텍스트로부터 이미지를 생성하는 작업입니다. 그 다음 stable diffusion의 특정 아키텍처에 대해 제안된 optimization 내용을 다루게 됩니다. text prompt로 추론을 수행할 때 원하는 설명을 넣게 되고, 추가 조건을 사용하여 reverse diffusion process가 진행됩니다. 즉, stable diffusion의 주요 구성 요소는 다음과 같이 textr embedder, noise generation, denoising neural network, image decoder가 포함됩니다. 

<img width="420" alt="image" src="https://github.com/eehoeskrap/PaperReview/assets/18207989/7d19a480-79f2-42bf-ac5b-efabf3ebfbe3">

### Specialized Kernels: Group Norm and GELU

Group normalization (GN)은 UNet 아키텍처에서 등장하게 됩니다. 이러한 정규화 기법은 feature map의 채널을 더 작은 group으로 나누고, 각 그룹을 독립적으로 정규화하여 GN에 배치 크기에 덜 의존하고, 광범위한 배치 크기 및 다른 네트워크 아키텍처에 더 적합하게 적용될 수 있도록 작동됩니다. 아래 식과 같이 각 특성값 $x_i$는 그룹의 평균값 및 분산 값으로 정규화 됩니다. 

<img width="412" alt="image" src="https://github.com/eehoeskrap/PaperReview/assets/18207989/39594ba5-3105-4d0a-8ef0-2f9cedc2f8bf">

GN이 이루어지게 되면 reshape, mean, variance, normalize의 연산이 차례대로 일어나게 되는데, 본 논문에서는 intermediate tensor 없이 single GPU command로 GPU Shader 형태의 고유한 커널을 설계했습니다.

Gaussian Error Linear Unit (GELU)는 BERT, GPT, ViT 모델 등에서 ReLU 대신 사용되는 activation function인데, 아래 식과 같이 multiplication, addition, Gaussian error function과 같은 많은 계산을 포함하게 됩니다.

<img width="378" alt="image" src="https://github.com/eehoeskrap/PaperReview/assets/18207989/8c9ab92e-4c5d-46d9-a018-147986f28690"> 

<img width="418" alt="image" src="https://github.com/eehoeskrap/PaperReview/assets/18207989/02868bd3-69b9-40a7-aac0-734d74939a96">

본 논문에서는 이 activation function을 split, multiplication 연산을 통해 GPU Shader를 구현하여 single draw call에서 실행할 수 있도록 했습니다. 

### Enhancing Attention Module Efficiency

Stable Diffusion의 text/image transformer는 text-to-image 생성 작업에 중요하게 작용하는 conditional distribution 모델링을 용이하게 하는데, 그럼에도 불구하고 self-cross-attention 메커니즘은 quadratic time, memory complexity로 인해 long sequence를 처리하기가 힘들다고 합니다. 그래서 이러한 computational bottleneck을 완화시키기 위해 2가지 최적화 방법을 제안합니다. 

#### Partially Fused Softmax

attention computation은 UNect 구조에서 중간 레이어에 들어가게 됩니다. 아래 시식과 같이 Query, Key, Value 행렬로 이루어져있고, 일반적으로 N, M은 d보다 훨씬 크다고 하네요. 

<img width="412" alt="image" src="https://github.com/eehoeskrap/PaperReview/assets/18207989/4d33ae7b-1354-42c9-ad32-8d6b6c326f72">

행렬에서 수행되는 softmax 연산은 2가지 단계로 나뉩니다. 첫번째는 reduction 연산, 두번째는 element-wise 연산입니다. reduction 연산 같은 경우 행렬 A의 각 행의 최대값과 수정된 지수 합 S의 계산을 의미합니다. 그 다음 벡터 L 및 S를 사용하여 행렬 A의 값을 정규화 하기 위해 element-wise 연산이 수행됩니다. 

<img width="419" alt="image" src="https://github.com/eehoeskrap/PaperReview/assets/18207989/b693ee1a-b307-4e8e-b97d-b7a762d9e2a6">

**Comment by eehoeskrap** [comment]
<img width="760" alt="image" src="https://github.com/eehoeskrap/PaperReview/assets/18207989/3e44eaa9-153e-4af9-9418-96ef1ea7567e">

**Comment by eehoeskrap** [comment]
GPU 인식을 최적화해 스마트폰에서도 Diffusion 모델 구축하는 방식
참고자료
https://discuss.pytorch.kr/t/230616-ai/1827
